### PR TITLE
Add grpc load balancing configs

### PR DESCRIPTION
**Pull request by ckiosidis**
# TL;DR
_Please replace this text with a description of what this PR accomplishes._

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/2510

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: clients/go/admin/client.go [hunk 1]
```diff
@@ -192,6 +192,11 @@ func initializeClients(ctx context.Context, cfg *Config, tokenCache pkce.TokenCa
 		opts = append(opts, authOpt)
 	}
 
+	if cfg.Balancer != "" {
```
Comment (honnix) [inline clients/go/admin/client.go:195]: nit: maybe `BalancerName`?

File: clients/go/admin/client_test.go [hunk 2]
```diff
@@ -276,3 +276,33 @@ func ExampleClientSetBuilder() {
 	_ = clientSet.AuthMetadataClient()
 	_ = clientSet.IdentityClient()
 }
+
+func TestBalancerConfigs(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ckiosidis) [reply clients/go/admin/client_test.go:280]: it won't throw errors unfortunately. If the balancer cannot be found it will default to pick_first

File: clients/go/admin/client.go (hunk 1 above)
Comment (ckiosidis) [reply clients/go/admin/client.go]: changed it to LoadBalancingPolicy, what do you think?

File: clients/go/admin/client_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (honnix) [reply clients/go/admin/client_test.go]: Maybe we should preload `roundrobin` package. Maybe `grpclb` as well but that seems to be deprecated.

[the remark this answers is not in the document]
Comment (ckiosidis) [reply clients/go/admin/client_test.go]: it didnt seem to be an issue in the tests, but i can add it in the next commit

**Comment by honnix** [comment]
> Update the comment as well? https://github.com/flyteorg/flyteidl/blob/master/clients/go/admin/config.go#L1-L4

This is not yet addressed.

File: clients/go/admin/client_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (honnix) [reply clients/go/admin/client_test.go]: I meant something like `import _ "google.golang.org/grpc/balancer/roundrobin"` in `client.go`.

[the remark this answers is not in the document]
Comment (ckiosidis) [reply clients/go/admin/client_test.go]: yeah i realised that i added it in the wrong file

File: clients/go/admin/client.go [hunk 3]
```diff
@@ -192,6 +195,13 @@ func initializeClients(ctx context.Context, cfg *Config, tokenCache pkce.TokenCa
 		opts = append(opts, authOpt)
 	}
 
+	if cfg.LoadBalancingPolicy != "" {
+		// The grpcClient will be created regardless of the value of the loadBalancingPolicy.
```
Comment (honnix) [inline clients/go/admin/client.go:199]: Let's not to repeat the comment?

**Review by honnix — APPROVED** [review]

**Comment by honnix** [comment]
Let's not try to close https://github.com/flyteorg/flyte/issues/2510 because datacatalog is not yet addressed. @ckiosidis

File: clients/go/admin/client.go [hunk 4]
```diff
@@ -192,6 +195,10 @@ func initializeClients(ctx context.Context, cfg *Config, tokenCache pkce.TokenCa
 		opts = append(opts, authOpt)
 	}
 
+	if cfg.LoadBalancingPolicy != "" {
+		opts = append(opts, grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingPolicy":"%v"}`, cfg.LoadBalancingPolicy)))
```
Comment (EngHabu) [inline clients/go/admin/client.go:199]: Should we expose the ServiceConfig as the config? instead of the load balancing policy?
Docs are here: https://github.com/grpc/grpc/blob/master/doc/service_config.md
Protobuf type is here: https://github.com/grpc/grpc-proto/blob/master/grpc/service_config/service_config.proto#L625

It looks like there is no current Go type (the existing one is still used by the code but marked as deprecated)... so we can make this config a string with links to the documentation.

wdyt @ckiosidis, @honnix ?

Comment (honnix) [reply clients/go/admin/client.go:199]: Right. That makes sense. Is there any other package we will need to preload?

Comment (ckiosidis) [reply clients/go/admin/client.go]: I pushed a change, let me know what you think @EngHabu

File: clients/go/admin/config.go [hunk 5]
```diff
@@ -68,6 +67,11 @@ type Config struct {
 	PkceConfig pkce.Config `json:"pkceConfig" pflag:",Config for Pkce authentication flow."`
 
 	Command []string `json:"command" pflag:",Command for external authentication token generation"`
+
+	// Set the ServiceConfig formatted as a json string https://github.com/grpc/grpc/blob/master/doc/service_config.md
+	// eg. { "loadBalancingConfig": [{ "round_robin":{} }], "methodConfig": [{ "name": [{ "service": "foo", "method": "bar" }, { "service": "baz" }], "timeout": "1.000000001s"}]}
+	// find the full schema here https://github.com/grpc/grpc-proto/blob/master/grpc/service_config/service_config.proto#L625
+	ServiceConfig string `json:"serviceConfig" pdflag:",Set the service config for the grpc client"`
```
Comment (honnix) [inline clients/go/admin/config.go:74]: Maybe `DefaultServiceConfig` reflects it better because it is the default if a service doesn't enforce override. ;)

File: clients/go/admin/client.go [hunk 6]
```diff
@@ -18,6 +18,9 @@ import (
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials"
 	"google.golang.org/grpc/health/grpc_health_v1"
+
+	_ "google.golang.org/grpc/balancer/grpclb"     //nolint
```
Comment (honnix) [inline clients/go/admin/client.go:22]: I think this is deprecated.

```suggestion
```

File: clients/go/admin/client.go [hunk 7]
```diff
@@ -192,6 +195,10 @@ func initializeClients(ctx context.Context, cfg *Config, tokenCache pkce.TokenCa
 		opts = append(opts, authOpt)
 	}
 
+	if cfg.ServiceConfig != "" {
+		opts = append(opts, grpc.WithDefaultServiceConfig(cfg.ServiceConfig))
```
Comment (honnix) [inline clients/go/admin/client.go:199]:
```suggestion
		opts = append(opts, grpc.WithDefaultServiceConfig(cfg.DefaultServiceConfig))
```

**Comment by ckiosidis** [comment]
fixed on https://github.com/flyteorg/flyteidl/pull/301
